[ConstraintSystem] Detect invalid initializer references early

xedin opened the pull request:
Currently invalid initializer references are detected and
diagnosed in solution application phase, but that's too
late because solver wouldn't have required information while
attempting to determine the best solution, which might result
in viable solutions being ignored in favour of incorrect ones e.g.

```swift
protocol P {
  init(value: Int)
}

class C {
  init(value: Int, _: String = "") {}
}

func make<T: P & C>(type: T.Type) -> T {
   return T.init(value: 0)
}
```

In this example `init` on `C` would be preferred since it
comes from the concrete type, but reference itself is invalid
because it's an attempt to construct class object using
metatype value via non-required initializer.

Situations like these should be recognized early and invalid
use like in case of `C.init` should be ranked lower or diagnosed
if that is the only possible solution.

Resolves: rdar://problem/47787705

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

xedin committed:
[ConstraintSystem] Detect invalid initializer references early

Currently invalid initializer references are detected and
diagnosed in solution application phase, but that's too
late because solver wouldn't have required information while
attempting to determine the best solution, which might result
in viable solutions being ignored in favour of incorrect ones e.g.

```swift
protocol P {
  init(value: Int)
}

class C {
  init(value: Int, _: String = "") {}
}

func make<T: P & C>(type: T.Type) -> T {
  return T.init(value: 0)
}
```

In this example `init` on `C` would be preferred since it
comes from the concrete type, but reference itself is invalid
because it's an attempt to construct class object using
metatype value via non-required initalizer.

Situations like these should be recognized early and invalid
use like in case of `C.init` should be ranked lower or diagnosed
if that is the only possible solution.

Resolves: rdar://problem/47787705

xedin committed:
[CSDiagnostics] Add diagnostics for two kinds of incorrect initializer reference

- Attempting to construct class object using metatype value via
  non-required initializer

- Referencing initializer of protocol metatype base

Both of the diagnostics are used by `AllowInvalidInitRef` fix.

xedin committed:
[CSApply] Remove obsolete `diagnoseInvalidDynamicConstructorReferences`

slavapestov commented on include/swift/AST/Expr.h:
```diff
@@ -476,7 +476,9 @@ class alignas(8) Expr {
-          [](const Expr *E) -> Type { return E->getType(); }) const;
+          [](const Expr *E) -> Type { return E->getType(); },
+      llvm::function_ref<bool(const Expr *)> isTypeReference =
+          [](const Expr *E) { return E->isTypeReference(); }) const;
```
Would it be cleaner if Expr::isTypeReference() took the getType callback, instead of passing two callbacks here?

xedin replied:
It already takes `getType` but it's insufficient because we need a way to retrieve declarations from constraint system for `UnresolvedDotExpr` and `UnresolvedMemberExpr`.

slavapestov replied:
Do you still need the Expr version then, and if so can you pass in an optional callback to look up overload choices too? It’s a complex piece of code and it’s suboptimal to repeat it twice I think

xedin replied:
Looks like `Expr::isStaticallyDerivedMetatype` is also used by SILGenApply and CodeCompletion besides CSApply.

xedin replied:
I'm going to commit the least ugly version of adding getDecl callback to `isTypeReference` I could come up with after I run tests locally.

slavapestov replied:
Awesome, thanks!

xedin replied:
Done. Added `getDecl` parameter to `isTypeRerefence` which defaults to `return nullptr` so I can use it in `ConstraintSystem::isTypeReference` to fetch declarations related to `UnresolvedDotExpr` and `UnresolvedMemberExpr`. WDYT?

slavapestov commented on lib/Sema/CSApply.cpp:
```diff
@@ -227,13 +227,57 @@ getImplicitMemberReferenceAccessSemantics(Expr *base, VarDecl *member,
+/// fully type-checked AST.
 bool ConstraintSystem::isTypeReference(const Expr *E) {
-  return E->isTypeReference([&](const Expr *E) -> Type { return getType(E); });
+  // If the result isn't a metatype, there's nothing else to do.
```
Why is this open-coded now?

slavapestov reviewed (commented):
Excellent!

xedin committed:
[AST] Add `getDecl` parameter to `Expr::isTypeRerefence`

So it could be extended to support not yet fully type-checked
AST when used by constraint system solver.

xedin commented:
@swift-ci please test

xedin commented:
@swift-ci please test source compatibility

xedin commented:
@shajrawi Test re-enabled by https://github.com/apple/swift/pull/22388 failed on iphonesimulator-i386

xedin commented:
@swift-ci please smoke test macOS platform

shajrawi commented:
This should fix it: https://github.com/apple/swift/pull/22414

xedin commented:
@swift-ci please test macOS platform

shajrawi commented:
`IRGen.witness_method.sil` - this time it is not my fault

xedin commented:
Looks like the fix for this failure just got merged in https://github.com/apple/swift/pull/22421

xedin commented:
@swift-ci please test macOS platform
